### PR TITLE
kola/tests: add updatepayload test

### PR DESCRIPTION
**Pull request by arithx**
Creates a new test `coreos.update.updatepayload` that mirrors the
functionality of `kola updatepayload` when an update payload is provided
via the `--update-payload` flag.

### CODE REVIEW

File: kola/tests/update/update.go [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+func updatepayload(c cluster.TestCluster) {
+	if kola.UpdatePayload == "" {
+		c.Skip("no update payload provided")
+	}
+
+	// create the actual test machine, the machine
+	// that is created by the test registration is
+	// used to host the omaha server if running on
+	// a non-qemu platform
+	m, err := c.NewMachine(conf.Ignition(`{"ignition":{"version": "2.1.0"}}`))
```
Comment (bgilbert) [inline kola/tests/update/update.go]: `c.NewMachine(nil)` is sufficient (though that will currently pass a 2.0.0 config, if that matters).

File: kola/tests/update/update.go [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+		c.Fatalf("invalidating USR-A failed: %s: %v: %s", out, err, stderr)
+	}
+
+	updateMachine(m, c)
+
+	checkUsrA(m, c)
+}
+
+func configureOmahaServer(c cluster.TestCluster) string {
+	if qc, ok := c.Cluster.(*qemu.Cluster); ok {
```
Comment (bgilbert) [inline kola/tests/update/update.go]: Is there value to either
1. simplifying the code by always using the dedicated omaha machine, or
2. not launching the omaha machine on qemu (which would require manually copying kolet)?

File: kola/tests/update/update.go [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+			c.Fatalf("copying update payload to omaha server: %v", err)
+		}
+
+		c.MustSSH(srv, fmt.Sprintf("sudo systemd-run --quiet ./kolet run %s Omaha", c.Name()))
+
+		return fmt.Sprintf("%s:34567", srv.PrivateIP())
+	}
+}
+
+func configureMachine(m platform.Machine, c cluster.TestCluster, addr string) {
```
Comment (bgilbert) [inline kola/tests/update/update.go]: Conventionally `c` before `m`.

File: kola/tests/update/update.go [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+6kvIuP+75SrMNeYAcU8PI1+bzLcAG3VN3jA78zeKALgynUNH50mxuiiU3DO4DZ+p
+5QIDAQAB
+-----END PUBLIC KEY-----
+EOF"`)
+
+	c.MustSSH(m, "sudo mount --bind /etc/coreos/update-payload-key.pub.pem /usr/share/update_engine/update-payload-key.pub.pem")
+
+	// disable reboot so the test has explicit control
+	c.MustSSH(m, "sudo systemctl mask locksmithd.service")
+	c.MustSSH(m, "sudo systemctl stop locksmithd.service")
```
Comment (bgilbert) [inline kola/tests/update/update.go]: The above can be combined into `mask --now`.

File: kola/tests/update/update.go [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+	c.MustSSH(m, "sudo systemctl reset-failed locksmithd.service")
+
+	c.MustSSH(m, "sudo systemctl restart update-engine.service")
+
+	c.MustSSH(m, "update_engine --v=10")
```
Comment (bgilbert) [inline kola/tests/update/update.go]: Left over from debugging?

File: kola/tests/update/update.go [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+	for _, p := range pairs {
+		spl := strings.SplitN(p, "=", 2)
+		if len(spl) == 2 {
+			m[spl[0]] = spl[1]
+		}
+	}
+	return m
+}
+
+func checkUsrA(m platform.Machine, c cluster.TestCluster) {
```
Comment (bgilbert) [inline kola/tests/update/update.go]: `misc` already has `assertBootedUsr()`.  Probably better to export that (maybe in a `util` package?) and reuse.

**Comment by arithx** [comment]
My initial vision was to produce a payload during manifest builds in jenkins and run at least the currently released version -> version being built. I do think that identifying specific versions for each channel would have some value as well.

File: kola/tests/update/update.go [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+			fmt.Printf("creating trivial omaha server: %v\n", err)
+		}
+
+		omahawrapper := local.OmahaWrapper{TrivialServer: omahaserver}
+
+		if err = omahawrapper.AddPackage("/updates/update.gz", "update.gz"); err != nil {
+			fmt.Printf("bad payload: %v", err)
+		}
+
+		go omahawrapper.Serve()
```
[the remark this answers is not in the document]
Comment (euank) [reply kola/tests/update/update.go]: Agreed; I'm not sure either of the goroutines are necessary.

File: kola/tests/update/update.go (hunk 2 above)
Comment (arithx) [reply kola/tests/update/update.go]: I'll play around with using a dedicated omaha machine for the QEMU cluster rather than the one provided via the `LocalCluster`.

File: kola/tests/update/update.go (hunk 6 above)
Comment (arithx) [reply kola/tests/update/update.go]: Thoughts on location for the containing `util` package? As this requires both `cluster.TestCluster` & `platform.Machine` it cannot live in `util/` or in `platform/util.go` because of circular imports. I'm not particularly partial to creating a `util` package inside of `kola/tests` either. Maybe inside of `kola/cluster`?

**Comment by euank** [comment]
I think it would be preferable if the test could dictate the version it's updating from.

e.g., it would be useful if we could have a test for `coreos.update.free-magic-grub` which updates a specific version of CL that has grub free-magic.

Right now, that test can't be written since the test would have to define a base CL version with a bad grub. As implemented, that base version is implied by the environment, so to run that test would be a matter of downloading that old version and running a regular update test.
Correct me if I'm misunderstanding how this would work.

I recognize that pragmatically it would be difficult to teach kola how to boot old versions on each platform and this implementation might be the best reasonable compromise, I just wanted to make sure that if we're choosing not to do the above after a conscious decision.

**Comment by arithx** [comment]
@euank you're understanding of the current implementation is correct; my thinking for now was that we would define an additional job inside of the jenkins-os that would take care of running `coreos.update.*` tests with a set of known verisons more manually e.g.:
```
for id in <ami_id1> <ami_id2>; do
  kola run -p aws --aws-ami <ami_id> coreos.update.*
done
```

I do agree that it would be nice to be able to pin versions to certain tests declaratively but the implementation on the runner side might not be something we'd want to add directly to `kola`. Maybe a seperate binary in mantle that's goal is directly targeting update testing from a set of known versions defined by the tests?

File: kola/tests/update/update.go (hunk 6 above)
Comment (bgilbert) [reply kola/tests/update/update.go]: I do think it makes sense to have a `kola.tests.util`, containing purely test code shared among multiple tests.  I'm open to other ideas though.

**Comment by arithx** [comment]
Update pushed

File: kola/tests/update/update.go [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+func ConfigureOmahaServer(c cluster.TestCluster, srv platform.Machine) string {
+	if kola.UpdatePayloadFile == "" {
+		c.Skip("no update payload provided")
+	}
+
+	if qc, ok := c.Cluster.(*qemu.Cluster); ok {
+		// An Omaha Server is built for all qemu clusters via the LocalCluster;
+		// Destroy it before setting up the omaha server on the machine created
+		// for the test.
+		qc.OmahaServer.Destroy()
```
Comment (bgilbert) [inline kola/tests/update/update.go]: Why is this necessary?  Also, is `OmahaServer` safe against multiple destructions?

Comment (arithx) [reply kola/tests/update/update.go]: I'll test without destroying this, I left the disabling code because previously it was overriding the `OmahaServer` I was setting up but that shouldn't be an issue anymore.

Comment (arithx) [reply kola/tests/update/update.go]: You were correct that this is no longer necessary. Now that the `OmahaServer` is being hosted on the original machine created by `kola` it is on a different IP than the host (`10.0.0.2` for the `kola` machine vs `10.0.0.1` for the host).

Pushed an update dropping this.

File: kola/tests/update/update.go [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+	in, err := os.Open(kola.UpdatePayloadFile)
+	if err != nil {
+		c.Fatalf("opening update payload: %v", err)
+	}
+	defer in.Close()
+	if err := platform.InstallFile(in, srv, "/updates/update.gz"); err != nil {
+		c.Fatalf("copying update payload to omaha server: %v", err)
+	}
+
+	c.MustSSH(srv, fmt.Sprintf("sudo systemd-run --quiet ./kolet run %s Omaha", c.Name()))
```
Comment (euank) [inline kola/tests/update/update.go]: There should also be a verification that the server is running before the address is returned, otherwise this is racing with the launching of the new machine.

File: kola/tests/update/update.go [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+
+	// create the actual test machine, the machine
+	// that is created by the test registration is
+	// used to host the omaha server
+	m, err := c.NewMachine(nil)
+	if err != nil {
+		c.Fatalf("creating test machine: %v", err)
+	}
+
+	ConfigureMachineForUpdate(c, m, addr)
```
Comment (euank) [inline kola/tests/update/update.go]: Is there a reason the machine isn't being configured with Ignition?

That'll make it easier to not race with update-engine for writing files (since the systemd unit that bindmounts over the pubkey can have a `Before=update-engine.service`).

I also think a CLC is more readable than the procedural ssh bits in this function.

File: kola/tests/update/update.go [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+	c.MustSSH(m, "sudo mount --bind /etc/coreos/update-payload-key.pub.pem /usr/share/update_engine/update-payload-key.pub.pem")
+
+	// disable reboot so the test has explicit control
+	c.MustSSH(m, "sudo systemctl mask --now locksmithd.service")
+	c.MustSSH(m, "sudo systemctl reset-failed locksmithd.service")
+
+	c.MustSSH(m, "sudo systemctl restart update-engine.service")
+}
+
+func UpdateMachine(c cluster.TestCluster, m platform.Machine) {
```
Comment (euank) [inline kola/tests/update/update.go]: doesn't need to be an exported function.

File: kola/tests/update/update.go [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+	err = util.WaitUntilReady(120*time.Second, 10*time.Second, func() (bool, error) {
+		envs, stderr, err := m.SSH("update_engine_client -status 2>/dev/null")
+		if err != nil {
+			return false, fmt.Errorf("checking status failed: %v: %s", err, stderr)
+		}
+
+		return splitNewlineEnv(string(envs))["CURRENT_OP"] == "UPDATE_STATUS_UPDATED_NEED_REBOOT", nil
+	})
+	if err != nil {
+		c.Fatalf("Updating machine: %v", err)
```
Comment (euank) [inline kola/tests/update/update.go]: A clearer message might be nice, e.g. `error waiting for STATUS_NEED_REBOOT: %v`

File: kola/tests/update/update.go (hunk 9 above)
Comment (bgilbert) [reply kola/tests/update/update.go]: In practice, is it likely for the kolet launch to lose the race with the instance launch?

Comment (euank) [reply kola/tests/update/update.go]: No. In reality we can probably leave this be, but I also think it would be nice to verify it's serving for the chance that kolet fails to launch correctly. `systemd-run` can quite easily eat errors and just spit out "yeah, I ran a unit" even if the unit instantly failed.

Comment (arithx) [reply kola/tests/update/update.go]: `update_engine` retries even in cases of no connectivity to the omaha server so this would only really validate that the omaha server eventually starts correctly. Is the thought to detect that failure case and fail the test earlier rather than failing when the machine fails to update and deciphering the logs?

File: kola/tests/update/update.go (hunk 10 above)
Comment (arithx) [reply kola/tests/update/update.go]: My initial thought for this was to leave the UserData completely up to the user so different tests could configure machines however they wanted and then these functions would do post-configuration.

Now that we're shipping an updated Ignition vendor I'll look into adding some new methods into `platform/conf` for combining configs and provide a function in this module that takes a custom userdata and adds in the requisite pieces for the update configuration.

File: kola/tests/update/update.go (hunk 11 above)
Comment (arithx) [reply kola/tests/update/update.go]: I saw this as more of a library function that other tests could utilize that was specific to updates. I can make this non-exported if we want to keep all update tests to this module.

File: kola/tests/update/update.go (hunk 10 above)
Comment (euank) [reply kola/tests/update/update.go]: After thinking about it again, it actually makes sense to not use Ignition since we may want to test updates on versions which predate ignition.

I think a comment to that effect on the function would be worth having.

File: kola/tests/update/update.go (hunk 9 above)
Comment (euank) [reply kola/tests/update/update.go]: That's a benefit, yeah. It's mostly about trying to have the function have a clear contract that it abides by.
Returning the address implies that it's already running unless there's documentation to the contrary.

File: kola/tests/update/update.go (hunk 11 above)
Comment (arithx) [reply kola/tests/update/update.go]: If we're going to make this function not exported should we also not export the `ConfigureOmahaServer` function?

Comment (euank) [reply kola/tests/update/update.go]: If it's a library function I wouldn't want it to live in the module. If it ends up being useful for another test, I'd rather it be moved and exported than have that other test import from another test module.

Each test module should be a leaf in the import graph; tests importing directly from other tests will eventually lead to cycles.

**Comment by arithx** [comment]
Update pushed
